web_src/js/svg.js

```diff
@@ -120,6 +119,45 @@ export const SvgIcon = {
     className: {type: String, default: ''},
   },
   render() {
-    return h('span', {innerHTML: svg(this.name, this.size, this.className)});
+    const svgStr = svgs[this.name];
+    if (!svgStr) throw new Error(`Unknown SVG icon: ${this.name}`);
+
+    // parse the SVG string to 2 parts
+    // * svgInnerHtml: the inner part of the SVG, will be used as the content of the <svg> VNode
+    // * svgOuter: the outer part of the SVG, including attributes
+    // the builtin SVG contents are clean, so it's safe to use `indexOf` to split the content:
+    // eg: <svg outer-attributes>${svgInnerHtml}</svg>
+    const p1 = svgStr.indexOf('>'), p2 = svgStr.lastIndexOf('<');
+    if (p1 === -1 || p2 === -1) throw new Error(`Invalid SVG icon: ${this.name}`);
+    const svgInnerHtml = svgStr.slice(p1 + 1, p2);
+    const svgOuterHtml = svgStr.slice(0, p1 + 1) + svgStr.slice(p2);
+    const svgDoc = parser.parseFromString(svgOuterHtml, 'image/svg+xml');
+    const svgOuter = svgDoc.firstChild;
```

Comment:
I do not think it's worth to parse the whole SVG string. We do not need to pay the cost for its inner.

Reply:
Where is the `parsed document` before you parse? When you parse, you spend time on its innter, but we never use its inner.

Reply:
OK, to be clear, let's say 2 approaches.

1. My approach: we only parse the Outer, never parse the inner. I think it's faster because we do not need to spend time on inner.
2. Your suggestion, parse the whole SVG string, including its outer and inner (slow path). Then we need to call "stringify(inner)" to get then the HTML string for VNode.innerHTML , that another slow path.

Do you really prefer your suggestion?

Reply:
All SVG are generated by your "generate-svg.js" , they are strictly clean , so that's absolute safe.

About "tests", I have no idea how to write a test for it ..... a normal SVG like `<svg>...</svg>`?

Reply:
Import `SvgIcon` into the test file and probably also Vue. Then make Vue render to a temporary element or string and test the output.

Reply:
Would prefer to test the whole thing. Not an expert on Vue, but something like this maybe:

```js
import {SvgIcon} from './svg.js';
import {createApp} from 'vue';

test('SvgIcon', () => {
  const root = document.createElement('div');
  createApp({data: () => SvgIcon('octicon-link', 24, 'class-name')}, root)
  expect(root.firstChild.nodeName).toEqual('SVG');
  expect(root.firstChild.height).toEqual(24);
  expect(root.firstChild.width).toEqual(24);
  expect(root.firstChild.className).toEqual('svg octicon-link');
});
```

Reply:
```
TypeError: Cannot read properties of null (reading 'nodeName')
```

`console.log(root.querySelectorAll('*').length);` says:  `0`

Reply:
Maybe we need https://test-utils.vuejs.org/guide/#what-is-vue-test-utils

Reply:
> Edit: Added `.mount()`

```
SvgIcon is not a function
TypeError: __vite_ssr_import_1__.SvgIcon is not a function
```

You have my fork's write access, feel free to edit this PR directly.

Reply:
Will check later, I feel like this should be possible without additional dependencies.

Reply:
Just for info, the reference example of vitest for vue uses test-utils, but we'll see if those are really necessary:

https://github.com/vitest-dev/vitest/blob/main/examples/vue/test/basic.test.ts

Reply:
Added 59f03a298516312ad7f2c15809f6064c6c307a4a

```
test('SvgIcon', () => {
  const root = document.createElement('div');
  createApp({render: () => h(SvgIcon, {name :'octicon-link', size: 24})}).mount(root);
  expect(root.firstChild.nodeName).toEqual('svg');
  expect(root.firstChild.getAttribute('width')).toEqual('24');
  expect(root.firstChild.classList.contains('octicon-link')).toBeTruthy();
}
```

Reply:
Nice, that should be it then.